src/main/java/org/folio/rest/impl/UsersAPI.java

```diff
@@ -228,40 +238,47 @@ public void postUsers(String lang, User entity, RoutingContext routingContext,
     }
   }
 
-  private void saveUser(User entity, Map<String, String> okapiHeaders, Handler<AsyncResult<Response>> asyncResultHandler,
+  private void saveUser(User entity, Map<String, String> okapiHeaders, PostgresClient pgClient, Handler<AsyncResult<Response>> asyncResultHandler,
       Context vertxContext) {
 
     Date now = new Date();
     entity.setCreatedDate(now);
     entity.setUpdatedDate(now);
-    PgUtil.post(TABLE_NAME_USERS, entity, okapiHeaders, vertxContext, PostUsersResponse.class, reply -> {
-      if (isDuplicateIdError(reply)) {
-        asyncResultHandler.handle(
-          succeededFuture(PostUsersResponse.respond422WithApplicationJson(
-            ValidationHelper.createValidationErrorMessage(
-              "id", entity.getId(),
-              "User with this id already exists"))));
-        return;
-      }
-      if (isDuplicateUsernameError(reply)) {
-        asyncResultHandler.handle(
-          succeededFuture(PostUsersResponse.respond422WithApplicationJson(
-            ValidationHelper.createValidationErrorMessage(
-              "username", entity.getUsername(),
-              "User with this username already exists"))));
-        return;
-      }
-      if (isDuplicateBarcodeError(reply)) {
-        asyncResultHandler.handle(
-          succeededFuture(PostUsersResponse.respond422WithApplicationJson(
-            ValidationHelper.createValidationErrorMessage(
-              "barcode", entity.getBarcode(),
-              "This barcode has already been taken"))));
-        return;
-      }
-      logger.debug("Save successful");
-      asyncResultHandler.handle(reply);
-    });
+    String userId = StringUtils.defaultIfBlank(entity.getId(), UUID.randomUUID().toString());
+
+    pgClient.withTrans(conn -> conn.saveAndReturnUpdatedEntity(TABLE_NAME_USERS, userId, entity.withId(userId))
+      .compose(user -> userOutboxService.saveUserOutboxLog(conn, user, UserEvent.Action.CREATE, okapiHeaders)
```

Comment:
How does the transaction get committed?

Is RMB doing something special to know that everything that happens on that connection should be in the same transaction?

Reply:
Under the hood, the withTrans mechanism is implemented to provide rollback functionality.